scene/animation/animation_tree.cpp

```diff
@@ -991,6 +997,7 @@ void AnimationTree::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::NODE_PATH, "anim_player", PROPERTY_HINT_NODE_PATH_VALID_TYPES, "AnimationPlayer"), "set_animation_player", "get_animation_player");
 
 	ADD_SIGNAL(MethodInfo(SNAME("animation_player_changed")));
+	ADD_SIGNAL(MethodInfo(SceneStringName(animation_node_notification), PropertyInfo(Variant::STRING_NAME, "animation_node_path"), PropertyInfo(Variant::INT, "what")));
```

Comment:
```suggestion
	ADD_SIGNAL(MethodInfo(SceneStringName(animation_node_notification), PropertyInfo(Variant::STRING_NAME, "animation_node_path"), PropertyInfo(Variant::STRING_NAME, "animation_node_class_name"), PropertyInfo(Variant::INT, "what"), PropertyInfo(Variant::PACKED_INT64_ARRAY, "sub_indices")));
```
Also I recommend adding `sub_indices` so that the transition source and destination can be passed.

If connect_deferred is used, they may not match when looking at the current Transition state by retrieving the AnimationTree parameters from within the signal. So it is safe to pass them as signal arguments.

Ideally, vararg should be used, but since signals cannot use vararg, so this is a compromise solution.

Reply:
I was thinking about this too, although initially I was leaning towards having an ultra generic `Variant extra` parameter since no obvious subset of state existed between all animation node types. `int[]` probably makes 
sense as the most specific type but I am very uncertain about effectively making adhoc marshaling/encoding a part of the signature and for the semantics to be completely disjoint between node types. It's also a bit of a slippery slope deciding what data gets to be included considering the premise is that the user needs to access a snapshot in time of the state. But then a more general data container would invite a situation similar to [this](https://sampruden.github.io/posts/godot-is-not-the-new-unity/) especially in light of this being a generic signal which is likely being immediately filtered on `what` in the callback.

In c# this question (as well as the enum problem) would be addressed by `EventArgs` but I think the only way to achieve something similar with the godot api would be by creating and passing a `RefCounted` subclass with each emit. This seems to me too expensive to opt all animation tree users into.

Currently I think the best thing would be to have these notifications in places where state is meaningful (ie `TRANSITION_STARTED` having the current and requested states both be populated) and rely on users treating the callback as the correct time to check/process the animation tree state. If there's such a case where the processing absolutely needs to be deferred, an intermediate callback could read the state and pass that to the deferred call.